### PR TITLE
Start

### PR DESCRIPTION
**Pull request by BoogieMonster1O1**
This pull request aims to include the simplest of simple modules present in fabric API.
This addresses the lack of adequate hooks and features in an environment with an over-engineered set of mappings and a lacking API.

This includes:
- [x] Item API (no registry API yet, but this can be used to add functionality to vanilla items)
- [x] Crash Report Info
- [x] Networking
- [x] Game rules
- [x] Commands
- [x] Key bindings
- [x] Lifecycle events

Notable changes:
- ~~Switch to [Chocohead's loom fork](https://github.com/Chocohead/Fabric-Loom) as it has support for pulling mappings off of Github~~ (Reverted)
- Change the license header to attribute Legacy Fabric
- Remove the `unfinished` folder and instead use a submodule
- Change the package name to `net.legacyfabric`
- Change `fabric.mod.json` contact metadata to match legacy fabric's
- Change the group to `net.legacyfabric.fabric-api`

### CODE REVIEW

**Comment by hYdos** [comment]
im all ok with all of that. the one request i have is you dont use chocoloom. im ok with the rest.

**Comment by hYdos** [comment]
i also ask you **dont** port networking v0 and instead port networking v1 as we dont need to port outdated apis.

**Comment by BoogieMonster1O1** [comment]
moving into a separate api

**Comment by BoogieMonster1O1** [comment]
yes i changeth my mind

**Comment by BoogieMonster1O1** [comment]
Fully tested and finished

File: .gitignore [hunk 1]
```diff
@@ -33,3 +33,7 @@ $RECYCLE.BIN/
 
 # Mac
 .DS_Store
+
+# Stuff
+reference114
```
Comment (hYdos) [inline .gitignore:38]: can you please not add random stuff to the gitignore.

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 2]
```diff
@@ -86,25 +80,22 @@ public static PlayChannelHandler unregisterGlobalReceiver(Identifier channelName
...
-	 * <p>For example, if you only register a receiver using this method when a {@linkplain ClientLoginNetworking#registerGlobalReceiver(Identifier, ClientLoginNetworking.LoginQueryRequestHandler)}
-	 * login query has been received, you should use {@link ClientPlayConnectionEvents#INIT} to register the channel handler.
+	 * Use {@link #unregisterReceiver(String)} to unregister the existing handler.
 	 *
 	 * @param channelName the id of the channel
 	 * @return false if a handler is already registered to the channel
 	 * @throws IllegalStateException if the client is not connected to a server
 	 * @see ClientPlayConnectionEvents#INIT
 	 */
-	public static boolean registerReceiver(Identifier channelName, PlayChannelHandler channelHandler) {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:107]: we could still force mods to use identifiers so atleast mods have a standard

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 3]
```diff
@@ -123,8 +114,7 @@ public static boolean registerReceiver(Identifier channelName, PlayChannelHandle
 	 * @return the previous handler, or {@code null} if no handler was bound to the channel
 	 * @throws IllegalStateException if the client is not connected to a server
 	 */
-	@Nullable
-	public static PlayChannelHandler unregisterReceiver(Identifier channelName) throws IllegalStateException {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:127]: same as above

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 4]
```diff
@@ -140,7 +130,7 @@ public static PlayChannelHandler unregisterReceiver(Identifier channelName) thro
 	 * @return All the channel names that the client can receive packets on
 	 * @throws IllegalStateException if the client is not connected to a server
 	 */
-	public static Set<Identifier> getReceived() throws IllegalStateException {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:143]: same as above

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 5]
```diff
@@ -156,7 +146,7 @@ public static PlayChannelHandler unregisterReceiver(Identifier channelName) thro
 	 * @return All the channel names the connected server declared the ability to receive a packets on
 	 * @throws IllegalStateException if the client is not connected to a server
 	 */
-	public static Set<Identifier> getSendable() throws IllegalStateException {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:159]: same as above

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 6]
```diff
@@ -189,7 +179,7 @@ public static boolean canSend(Identifier channelName) throws IllegalArgumentExce
 	 * @param buf the packet byte buf which represents the payload of the packet
 	 * @return a new packet
 	 */
-	public static Packet<?> createC2SPacket(Identifier channelName, PacketByteBuf buf) {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:192]: same as above

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 7]
```diff
@@ -173,7 +163,7 @@ public static PlayChannelHandler unregisterReceiver(Identifier channelName) thro
 	 * @return True if the connected server has declared the ability to receive a packet on the specified channel.
 	 * False if the client is not in game.
 	 */
-	public static boolean canSend(Identifier channelName) throws IllegalArgumentException {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:176]: same as above

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java [hunk 8]
```diff
@@ -218,7 +208,7 @@ public static PacketSender getSender() throws IllegalStateException {
 	 * @param buf the payload of the packet
 	 * @throws IllegalStateException if the client is not connected to a server
 	 */
-	public static void send(Identifier channelName, PacketByteBuf buf) throws IllegalStateException {
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/client/networking/v1/ClientPlayNetworking.java:221]: I think you get the point

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/networking/v1/EntityTrackingEvents.java [hunk 9]
```diff
@@ -14,22 +15,23 @@
...
-
 /**
  * Events related to a tracking entities within a player's view distance.
  */
 public final class EntityTrackingEvents {
 	/**
 	 * An event that is called before player starts tracking an entity.
 	 * Typically this occurs when an entity enters a client's view distance.
-	 * This event is called before the player's client is sent the entity's {@link Entity#createSpawnPacket() spawn packet}.
+	 * This event is called before the player's client is sent the entity's {@link EntityTrackerEntry#method_33558() spawn packet}.
```
Comment (hYdos) [inline fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/networking/v1/EntityTrackingEvents.java:34]: someone should map this

**Review by hYdos — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (BoogieMonster1O1) [reply .gitignore:38]: sorry i need that one

Comment (DeadlyMC) [reply .gitignore:38]: We can't be adding random stuff that exists on your system to the github gitignore

Comment (BoogieMonster1O1) [reply .gitignore:38]: thats the folder i used for the 1.14 fabric api source code
its fine cmon

Comment (DeadlyMC) [reply .gitignore:38]: No its not, just move your source code to a different directory.

Comment (BoogieMonster1O1) [reply .gitignore]: irritat

File: fabric-crash-report-info-v1/src/main/java/net/legacyfabric/fabric/mixin/crashreport/CrashReportMixin.java [hunk 10]
```diff
@@ -32,13 +34,14 @@
 import net.fabricmc.loader.api.ModContainer;
 
 @Mixin(CrashReport.class)
-public abstract class MixinCrashReport {
+public class CrashReportMixin {
```
Comment (DeadlyMC) [inline fabric-crash-report-info-v1/src/main/java/net/legacyfabric/fabric/mixin/crashreport/CrashReportMixin.java:37]: Why did you remove the `abstract` ?  Its a good practice to make your mixins `abstract` because it means you can easily add abstract members (eg. shadow methods) without the compiler complaining.
it also means you can't accidentally do `new CrashReportMixin` anywhere else in your code by mistake.

File: fabric-item-api-v1/src/main/resources/fabric.mod.json [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  "license": "Apache-2.0",
+  "icon": "assets/fabric-item-api-v1/icon.png",
+  "contact": {
+    "homepage": "https://legacyfabric.net",
+    "irc": "irc://irc.esper.net:6667/legacyfabric",
+    "issues": "https://github.com/Legacy-Fabric/fabric/issues",
+    "sources": "https://github.com/Legacy-Fabric/fabric"
+  },
+  "authors": [
+    "FabricMC"
```
Comment (DeadlyMC) [inline fabric-item-api-v1/src/main/resources/fabric.mod.json:16]: I think you missed this.

File: fabric-keybinding-api-v1/src/main/java/net/legacyfabric/fabric/mixin/keybinding/KeyBindingAccessor.java [hunk 12]
```diff
@@ -25,8 +26,8 @@
 
 @Mixin(KeyBinding.class)
 public interface KeyBindingAccessor {
-	@Accessor("categoryOrderMap")
-	static Map<String, Integer> fabric_getCategoryMap() {
-		throw new AssertionError();
+	@Accessor
+	static Map<String, Integer> getCATEGORIES() {
```
Comment (DeadlyMC) [inline fabric-keybinding-api-v1/src/main/java/net/legacyfabric/fabric/mixin/keybinding/KeyBindingAccessor.java:30]: Why not pass the field to the annotation and have a not so cursed method name?

Comment (BoogieMonster1O1) [reply fabric-keybinding-api-v1/src/main/java/net/legacyfabric/fabric/mixin/keybinding/KeyBindingAccessor.java]: why so many reviews on code style rather than code

Comment (DeadlyMC) [reply fabric-keybinding-api-v1/src/main/java/net/legacyfabric/fabric/mixin/keybinding/KeyBindingAccessor.java]: cause yes

**Comment by DeadlyMC** [comment]
I'm happy with everything in this pr except for the networking api, i would like to have a chat about it with everyone over at legacyfabricord.

**Comment by DeadlyMC** [comment]
Could you please remove your git submodule, thx

File: build.gradle [hunk 13]
```diff
@@ -123,26 +118,22 @@ allprojects {
 		maven { url "https://jitpack.io" }
 	}
 
+	dependencies {
+		minecraft "com.mojang:minecraft:$Globals.mcVersion"
+		mappings "net.fabricmc:yarn:$Globals.yarnVersion:v2"
+		modCompile "com.github.shedaniel:fabric-loader:${Globals.loaderVersion}"
```
Comment (phillyfox) [inline build.gradle:124]: `modImplementation`

File: build.gradle [hunk 14]
```diff
@@ -339,67 +312,3 @@ dependencies {
...
-
-		afterEvaluate {
-			uploadTask.dependsOn("remapJar")
-		}
-	}
-
-	options {
-		forgeGradleIntegration = false
-	}
-}
```
Comment (phillyfox) [inline build.gradle:367]: we're still gonna upload to curse, so re add this part

File: fabric-crash-report-info-v1/src/main/resources/fabric.mod.json [hunk 15]
```diff
@@ -1,23 +1,20 @@
...
+    "irc": "irc://irc.esper.net:6667/legacyfabric",
+    "issues": "https://github.com/Legacy-Fabric/fabric/issues",
+    "sources": "https://github.com/Legacy-Fabric/fabric"
   },
   "authors": [
-    "FabricMC"
+    "Legacy Fabric"
   ],
-  "depends": {
-    "fabricloader": ">=0.4.0"
```
Comment (phillyfox) [inline fabric-crash-report-info-v1/src/main/resources/fabric.mod.json:19]: why?

File: fabric-gamerule-api-v1/build.gradle [hunk 16]
```diff
@@ -0,0 +1,7 @@
+archivesBaseName = "fabric-gamerule-api-v1"
+version = getSubprojectVersion(project, "1.0.0")
+
+moduleDependencies(project, [
+		'fabric-api-base',
+		'fabric-lifecycle-events-v1'
```
Comment (phillyfox) [inline fabric-gamerule-api-v1/build.gradle:6]: unused dependency

File: fabric-item-api-v1/src/main/java/net/legacyfabric/fabric/api/item/v1/Enchantable.java [hunk 17]
```diff
@@ -14,13 +15,25 @@
...
 
-import net.minecraft.client.render.VertexConsumer;
-import net.minecraft.client.render.model.BakedModel;
-import net.minecraft.client.util.math.MatrixStack;
+import net.minecraft.item.Item;
 import net.minecraft.item.ItemStack;
 
-public interface AccessItemRenderer {
-	void fabric_renderBakedItemModel(BakedModel model, ItemStack stack, int light, int overlay, MatrixStack matrixStack, VertexConsumer buffer);
+public interface Enchantable {
```
Comment (phillyfox) [inline fabric-item-api-v1/src/main/java/net/legacyfabric/fabric/api/item/v1/Enchantable.java:23]: This name is misleading
Maybe `CustomEnchantabilityProvider`?

File: fabric-keybinding-api-v1/build.gradle [hunk 18]
```diff
@@ -0,0 +1,10 @@
+archivesBaseName = "fabric-keybinding-api-v1"
+version = getSubprojectVersion(project, "1.0.0")
+
+dependencies {
+	testmodImplementation project(path: ":fabric-lifecycle-events-v1", configuration: 'dev')
+}
+
+moduleDependencies(project, [
+		'fabric-api-base'
```
Comment (phillyfox) [inline fabric-keybinding-api-v1/build.gradle:9]: Unused dependency

File: fabric-lifecycle-events-v1/src/main/java/net/legacyfabric/fabric/api/event/lifecycle/v1/ServerTickEvents.java [hunk 19]
```diff
@@ -33,7 +34,7 @@ private ServerTickEvents() {
 	public static final Event<StartTick> START_SERVER_TICK = EventFactory.createArrayBacked(StartTick.class, callbacks -> server -> {
 		if (EventFactory.isProfilingEnabled()) {
 			final Profiler profiler = server.profiler;
-			profiler.push("fabricStartServerTick");
+			profiler.swap("fabricStartServerTick");
```
Comment (phillyfox) [inline fabric-lifecycle-events-v1/src/main/java/net/legacyfabric/fabric/api/event/lifecycle/v1/ServerTickEvents.java:37]: why?

**Review by phillyfox — CHANGES_REQUESTED** [review]

File: fabric-lifecycle-events-v1/src/main/java/net/legacyfabric/fabric/api/event/lifecycle/v1/ServerTickEvents.java (hunk 19 above)
Comment (BoogieMonster1O1) [reply fabric-lifecycle-events-v1/src/main/java/net/legacyfabric/fabric/api/event/lifecycle/v1/ServerTickEvents.java:37]: because the invoker injects right before a profiler call

File: fabric-crash-report-info-v1/src/main/resources/fabric.mod.json (hunk 15 above)
Comment (BoogieMonster1O1) [reply fabric-crash-report-info-v1/src/main/resources/fabric.mod.json:19]: because `patch-1` is not semver and it breaks comparison

**Review by hYdos — APPROVED** [review]

File: fabric-item-api-v1/src/main/java/net/legacyfabric/fabric/api/item/v1/Enchantable.java (hunk 17 above)
Comment (hYdos) [reply fabric-item-api-v1/src/main/java/net/legacyfabric/fabric/api/item/v1/Enchantable.java:23]: i dont see how its misleading. can you elaborate?

Comment (BoogieMonster1O1) [reply fabric-item-api-v1/src/main/java/net/legacyfabric/fabric/api/item/v1/Enchantable.java:23]: its because you dont need to implement this for the item to be enchantable.
@DeadlyMC suggest a name please

File: fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/networking/v1/EntityTrackingEvents.java (hunk 9 above)
Comment (DeadlyMC) [reply fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/networking/v1/EntityTrackingEvents.java:34]: noted

Comment (DeadlyMC) [reply fabric-networking-api-v1/src/main/java/net/legacyfabric/fabric/api/networking/v1/EntityTrackingEvents.java:34]: Mapped this, @BoogieMonster1O1 make the change

**Review by phillyfox — APPROVED** [review]
